### ISSUE TITLE
IDE0019 suggests C# 7.1-compatible fixes on a 7.0 project

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#26485]
I have a project that has its &quot;Language version&quot; set to &quot;C# latest major version (default)&quot;, which means the code is compiled in C# 7.0 mode.

In one of the files in this project, I have the following code:

```
struct MyStructType
{
   ...
}

...

object foo;

...

var val = foo as MyStructType; // IDE0019
if (val != null)
{
    ...
}
```

IDE0019 is triggered on the &quot;foo as MyStructType&quot; line, and it suggests to rewrite that code and the following if statement as:

```
if (foo is MyStructType val )
{
    ...
}
```

But this construct (pattern matching against struct types) does not work in C# 7.0 due to an oversight that was fixed in 7.1

Hence, IDE0019 should not recommend this refactoring unless the file in question is contained in a project that has its language version set to C# 7.1 or higher.

Furthermore, the documentation at https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching should be updated to note that it will only work in C# 7.1 or higher, because the definitions of the `Rectangle` and `Triangle` types as structs means that none of the example code will compile in any project that uses C# 7.0 (which is also the default version for new projects).

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/238468/ide0019-suggests-c-71-compatible-fixes-on-a-70-pro.html
VSTS ticketId: 604807_
_These are the original issue comments:_

Ian Kemp on 4/23/2018, 02:08 AM (5 days ago): <p>Dear god but your markdown parser mangles code.</p>

_These are the original issue solutions:_
(no solutions)

**Comment by Neme12** [comment]
Hi, regarding this example:
```c#
struct MyStructType
{
}

class C
{
    void M()
    {
        object foo = null;

        var val = foo as MyStructType; // IDE0019
        if (val != null)
        {
        }
    }
}
```
this code seems to be incorrect because the `as` operator doesn't work with structs. Indeed `val` can never be null as a struct. Are you sure you haven't missed something? Could you provide a full working sample? Thanks.

On the other hand, I'm not aware of any limitations using pattern matching with structs in C# 7.0. This code should be correct and compiles just fine:
```c#
struct MyStructType
{
}

class C
{
    void M()
    {
        object foo = null;

        if (foo is MyStructType val)
        {
        }
    }
}
```
Maybe I misunderstood the question. Can you please clarify what's wrong here? Thanks!
